src/qspi.rs

```diff
@@ -0,0 +1,230 @@
+            _ => return Err(QspiError::InvalidClock),
+        };
+
+        // Disable QUADSPI before configuring it.
+        regs.cr.write(|w| {
+            w.en().clear_bit()
+        });
+
+        // Configure the FSIZE to maximum. It appears that even when addressing is not used, the
+        // flash size violation may still trigger.
+        regs.dcr.write(|w| unsafe {
+            w.fsize().bits(0x1F)
+        });
+
+        // Clear all pending flags.
+        regs.fcr.write(|w| {
+            w
+            .ctof().set_bit()
+            .csmf().set_bit()
+            .ctcf().set_bit()
+            .ctef().set_bit()
+        });
+
+        // Configure the communication method for QSPI.
+        regs.ccr.write(|w| unsafe {
+            w
+            .fmode().bits(0)
+            .dmode().bits(0b01)
+            .imode().bits(0b01)
+        });
+
+        let spi_frequency = frequency.into().0;
+        let divisor = match spi_kernel_ck / spi_frequency {
+            1..=255 => {
+                spi_kernel_ck / spi_frequency - 1
+            },
+            _ => {
+                return Err(QspiError::InvalidFrequency);
+            },
+        };
+
+        // Write the prescaler and select flash bank 2 - flash bank 1 is currently unsupported.
+        regs.cr.write(|w| unsafe {
+            w.prescaler().bits(divisor as u8)
+             .fsel().set_bit()
+        });
+
+        Ok(Qspi{rb: regs})
+    }
+
+    pub fn is_busy(&self) -> bool {
+        self.rb.sr.read().busy().bit_is_set()
+    }
+
+    fn get_clock(ccdr: &Ccdr) -> Option<Hertz> {
+        match ccdr.rb.d1ccipr.read().qspisel().variant() {
+            stm32::rcc::d1ccipr::QSPISEL_A::RCC_HCLK3 => Some(ccdr.clocks.hclk()),
+            stm32::rcc::d1ccipr::QSPISEL_A::PLL1_Q => ccdr.clocks.pll1_q_ck(),
+            stm32::rcc::d1ccipr::QSPISEL_A::PLL2_R => ccdr.clocks.pll2_r_ck(),
+            stm32::rcc::d1ccipr::QSPISEL_A::PER => ccdr.clocks.per_ck(),
+        }
+    }
+
+    pub fn configure_mode(&mut self, mode: QspiMode) -> Result<(), QspiError> {
+        if self.rb.sr.read().busy().bit_is_set() {
+            return Err(QspiError::Busy);
+        }
+
+        match mode {
+            QspiMode::OneBit => {
+                self.rb.ccr.modify(|_, w| unsafe {
+                    w.imode().bits(0b01)
+                     .dmode().bits(0b01)
+                });
+
+                Ok(())
+            },
+            QspiMode::FourBit => {
+                self.rb.ccr.modify(|_, w| unsafe {
+                    w.imode().bits(0b11)
+                     .dmode().bits(0b11)
+                });
+
+                Ok(())
+            },
+            _ => Err(QspiError::Unsupported)
+        }
+    }
+
+    pub fn write(&mut self, addr: u8, data: &[u8]) -> Result<(), QspiError> {
+        if self.is_busy() {
+            return Err(QspiError::Busy);
+        }
+
+        // Clear the transfer complete flag.
+        self.rb.fcr.modify(|_ ,w| w.ctcf().set_bit());
```

Comment:
Why do you `ctcf()` before and after each read and write? Should be only after obtaining and acknowledging a set `tcf()`, i.e. after each read/write.

Reply:
This is done just to ensure that the transfer complete flag is clear before starting the transaction at all. This prevents a latent, uncleared flag (e.g. caused by a user using the QSPI peripheral before instantiation) from causing a transfer to immediately be detected as complete. This is just ensuring that we have the QSPI peripheral in a "known-good" state and is redundant in 99% of cases, but it doesn't cause any harm.